src/quantize_fp16.cpp

```diff
@@ -0,0 +1,165 @@
+        // process all inputs, if input is a fp32 or fp64, convert it
+        // to a fp16 by adding a convert operator.
+        auto inputs = ins->inputs();
+        std::vector<instruction_ref> converted_inputs;
+        for(auto input : inputs)
+        {
+            auto s = input->get_shape();
+            if(s.type() != shape::float_type and s.type() != shape::double_type)
+            {
+                converted_inputs.push_back(input);
+                continue;
+            }
+
+            // if the input is a parameter of a subgraph
+            instruction_ref input_fp16{};
+            if(input->name() == "@param")
+            {
+                auto in_outs = input->outputs();
+                auto it      = std::find_if(in_outs.begin(), in_outs.end(), [](auto o) {
+                    return (o->name() == "convert" and o->get_shape().type() == shape::half_type);
+                });
+                assert(it != in_outs.end());
+                input_fp16 = *it;
+                converted_inputs.push_back(input_fp16);
+            }
+            // if the input is a convert operator, uses its input
+            // as its current input
+            else if(input->name() == "convert" and
+                    input->inputs().front()->get_shape().type() == shape::half_type)
+            {
+                input_fp16 = input->inputs().front();
+                converted_inputs.push_back(input_fp16);
+            }
+            else
+            {
+                input_fp16 = insert_quant_ins(m, ins, input, shape::half_type, map_fp16);
+                converted_inputs.push_back(input_fp16);
+            }
+        }
+
+        auto mod_inputs = ins->module_inputs();
+        for(auto*& smod : mod_inputs)
+        {
+            quantize_module(*smod, ins_names, map_fp16, true);
+        }
+
+        auto op        = ins->get_operator();
+        auto ins_shape = compute_shape(op, converted_inputs, mod_inputs);
+        if(ins_shape != orig_shape)
+        {
+            // tuple type, followed by get_tuple_elem
+            if(ins_shape.type() == shape::tuple_type)
+            {
+                auto outputs = ins->outputs();
+                for(auto out : outputs)
+                {
+                    auto out1 = m.insert_instruction(
+                        std::next(out),
+                        make_op("convert", {{"target_type", out->get_shape().type()}}),
+                        out);
+                    m.replace_instruction(out, out1);
+                }
+            }
+            else
+            {
+                // check the dead code case to avoid assert
+                auto ins_orig_shape = m.insert_instruction(
+                    std::next(ins), make_op("convert", {{"target_type", orig_shape.type()}}), ins);
+                m.replace_instruction(ins, ins_orig_shape);
+            }
+        }
+        m.replace_instruction(ins, op, converted_inputs, mod_inputs);
+    }
+}
```

Comment:
This just seems more complicated then it needs to be. We already have passes that can remove redundant convert ops(and if its not sufficient with modules then we can look at expanding the optimization passes). So the quantize can just focus on inserting the ops needed to quantize a single op. Basically, insert a convert back to the original type, and then insert converts to fp16 for each of the inputs that are floats:

```cpp
for(auto ins : iterator_for(m))
{
    if (contains({"@return", "convert"}, ins->name()))
        continue;
    if (ins->intputs().empty())
        continue;

    auto r = m.insert_instruction(std::next(ins), make_op("convert", {{"target_type", s.type()}}));
    m.replace_instruction(ins, r);

    auto inputs = ins->inputs();
    std::transform(inputs.begin(), inputs.end(), inputs.begin(), [&](auto input) {
        auto s = input->get_shape();
        if(s.type() != shape::float_type and s.type() != shape::double_type)
            return input;
        return m.insert_instruction(ins, make_op("convert", {{"target_type", shape::half_type}}), input);
    });
    m.replace_instruction(ins, ins->get_operator(), inputs);
}
```

We always skip `@return`, `convert`(regardless of the target type), and operators with no inputs.

Reply:
This can be done using just a module pass. And there is no need to traverse modules or keep track of instruction in the parent modules.

Reply:
This is true for the main module, but for submodules, I think we need to convert "return" and "params" to fp16.

Reply:
We should skip quantizing ops that have submodules for now(we can still quantize the submodule) as I think there is more work to make it work correctly. 

If we change an input arg to be fp16 and the corresponding parameter is still fp32 then we have created an invalid program. We ultimately need a way to change the arguments and then update the corresponding parameters in the submodules together, but let's not tackle this for now.